src/Stratis.Bitcoin.Tests/Base/ConsensusManagerBehaviorTests.cs

```diff
@@ -0,0 +1,278 @@
+using System.Threading.Tasks;
+using Moq;
+using NBitcoin;
+using Stratis.Bitcoin.Base;
+using Stratis.Bitcoin.Configuration.Logging;
+using Stratis.Bitcoin.Connection;
+using Stratis.Bitcoin.Consensus;
+using Stratis.Bitcoin.Interfaces;
+using Stratis.Bitcoin.P2P.Peer;
+using Stratis.Bitcoin.P2P.Protocol;
+using Stratis.Bitcoin.P2P.Protocol.Payloads;
+using Stratis.Bitcoin.Tests.Common;
+using Stratis.Bitcoin.Utilities;
+using Xunit;
+
+namespace Stratis.Bitcoin.Tests.Base
+{
+    public class ConsensusManagerBehaviorTests
+    {
+        private bool IsIBD = false;
+
+        private AsyncExecutionEvent<INetworkPeer, NetworkPeerState> stateChanged;
+        private AsyncExecutionEvent<INetworkPeer, IncomingMessage> messageReceived;
+
+        private Mock<INetworkPeer> peerMock;
+
+        private readonly List<ChainedHeader> headers;
+
+        private readonly ExtendedLoggerFactory loggerFactory;
+
+        /// <summary>How many times behavior called the <see cref="ConsensusManager.HeadersPresented"/>.</summary>
+        private int headersPresentedCalledTimes;
+
+        public ConsensusManagerBehaviorTests()
+        {
+            this.loggerFactory = new ExtendedLoggerFactory();
+            this.loggerFactory.AddConsoleWithFilters();
+
+            this.headers = ChainedHeadersHelper.CreateConsecutiveHeaders(100);
+        }
+
+        public class Counter
+        {
+            public int Value = 0;
+        }
+
+        private Mock<INetworkPeer> CreatePeerMock()
+        {
+            var peer = new Mock<INetworkPeer>();
+
+            var connection = new NetworkPeerConnection(Network.StratisMain, peer.Object, new TcpClient(), 0, (message, token) => Task.CompletedTask,
+                new DateTimeProvider(), this.loggerFactory, new PayloadProvider());
+
+            peer.SetupGet(networkPeer => networkPeer.Connection).Returns(connection);
+
+            var connectionParameters = new NetworkPeerConnectionParameters();
+            VersionPayload version = connectionParameters.CreateVersion(new IPEndPoint(1, 1), Network.StratisMain, new DateTimeProvider().GetTimeOffset());
+            version.Services = NetworkPeerServices.Network;
+
+            peer.SetupGet(x => x.PeerVersion).Returns(version);
+            peer.SetupGet(x => x.State).Returns(NetworkPeerState.HandShaked);
+            peer.SetupGet(x => x.MessageReceived).Returns(new AsyncExecutionEvent<INetworkPeer, IncomingMessage>());
+
+            this.stateChanged = new AsyncExecutionEvent<INetworkPeer, NetworkPeerState>();
+            this.messageReceived = new AsyncExecutionEvent<INetworkPeer, IncomingMessage>();
+
+            peer.Setup(x => x.StateChanged).Returns(() => this.stateChanged);
+            peer.Setup(x => x.MessageReceived).Returns(() => this.messageReceived);
+
+            return peer;
+        }
+
+        private ConsensusManagerBehavior CreateAndAttachBehavior(ChainedHeader consensusTip, out Counter getHeadersPayloadSentTimes,
```

Comment:
Very hardly readable without comments of the arguments

Reply:
Fixed